sdk-extensions/baggage-processor/src/main/java/io/opentelemetry/sdk/extension/trace/baggage/BaggageSpanProcessorBuilder.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.extension.trace.baggage;
+
+import io.opentelemetry.sdk.metrics.internal.view.StringPredicates;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Builder for {@link BaggageSpanProcessor}.
+ *
+ * <p>Allows configurable rules for what baggage to include and when to include it.
+ */
+public class BaggageSpanProcessorBuilder {
+  private final SpanProcessor next;
```

Comment:
This might be an issue with the span processor API in general but while the pattern of wrapping could probably have been a better one for ensuring ordering, I don't think it's idiomatic for our SDK's current design. We'd expect the user to make sure this processor is added first, and not have wrapping

Reply:
Interesting. I followed the pattern of the AsyncProcessor here.  Do we have any instances of pipelining processors outside of these two?

Reply:
AsyncProcessor, BatchSpanProcessor, SimpleSpanProcessor are just exporter wrappers so they're a bit different from something that works with the actual span. A notable issue with this approach for baggage is that while intuitively, you'd expect modifications from the processor to only be available in `next` - in actuality, the span is mutated so other span processors in the pipeline, not just the `next`, will see the mutated span.

I think the only other non-exporter span processor we have is zpages

https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/zpages/src/main/java/io/opentelemetry/sdk/extension/zpages/TracezSpanProcessor.java

Reply:
AsyncProcessor does not wrap Exporter, but processor, see: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/async-processor/src/main/java/io/opentelemetry/sdk/extension/trace/export/DisruptorAsyncSpanProcessor.java#L72

It just looks funky because it hides the processor behind disruptor event queue fluffier.

Reply:
Oops didn't notice async doesn't explicitly take exporter, I think the only use case anyone expects with it is exporter so I guess it always requires a SimpleSpanProcessor in between...

Either way my comment still stands though, especially that a span mutating processor has no reason to wrap another

Reply:
In practice, we always recommend using a batch span processor.  For this specific processor I don't think preventing the usage of BSP is a great idea.

If a span-mutating process doesnt allow further processing, doesn't that mean it needs to have all functionality user could want?  Shouldn't all non-final processors wrap another?

Reply:
The SDK has `addSpanProcessor` and allows multiple processors that execute in order. So it doesn't prevent BatchSpanProcessor (phew since it's basically required to export). That's basically what makes zpages possible.

It means a batch span processor added to the SDK after baggage processor will be affected by the baggage processor regardless of if it's next or not. So I think the general idiom is for non-exporting to not wrap anything, and async's API is probably a bug (it's still alpha, it's probably around time to delete that module since I haven't heard any demand for it yet).

Reply:
The expressiveness of the export pipeline not being great came up around release but we decided it's too much to change the spec at that point. I was interested in getting rid of BSP type of exporting processors entirely. But for now we just have a list of processors in order - it doesn't make sense for a mutating processor to wrap since there is no isolation and non-wrapped will be effected too, defeating the expected meaning of wrapping.